caddy: add all addons

compile caddy with all addons listed on the [download page](https://caddyserver.com/download).

> Only middleware handlers which are invoked from the Caddyfile will be chained in, so small Caddyfiles are very fast and efficient.

source: Directive section on https://caddyserver.com/docs/caddyfile

a later improvement could be to select just the addons you want like vim_configurable.

@angus-g

@ntninja I've tried this, but it's not working for me. I recreated the problem in a flake for easier debugging:

https://github.com/crabdancing/nixos-caddy-with-plugins

It says:

```
warning: creating lock file '/home/ada/projects/nixos-caddy-with-plugins/flake.lock'
error: builder for '/nix/store/apjbwk230vf97ckwcl8rqa7dd7lqln3q-caddy-2.7.5.drv' failed with exit code 1;
       last 10 log lines:
       > Running phase: unpackPhase
       > unpacking source archive /nix/store/6qhv0b3ripxgjyw8jhibqmn1i3r91bzs-source
       > source root is source
       > Running phase: patchPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > Running phase: buildPhase
       > Building subPackage ./cmd/caddy
       > # github.com/caddyserver/caddy/v2/modules/caddytls
       > modules/caddytls/automation.go:254:18: cannot use func(name string) error {…} (value of type func(name string) error) as func(ctx context.Context, name string) error value in struct literal
       For full logs, run 'nix log /nix/store/apjbwk230vf97ckwcl8rqa7dd7lqln3q-caddy-2.7.5.drv'.
```

Does anyone have any ideas?

--- a/nixos/doc/manual/release-notes/rl-2505.section.md
+++ b/nixos/doc/manual/release-notes/rl-2505.section.md
@@ -160,6 +160,21 @@
 
 - `bind.cacheNetworks` now only controls access for recursive queries, where it previously controlled access for all queries.
 
+- Caddy can now be built with plugins by using `caddy.withPlugins`, a `passthru` function that accepts an attribute set as a parameter. The `plugins` argument represents a list of Caddy plugins, with each Caddy plugin being a versioned module. The `hash` argument represents the `vendorHash` of the resulting Caddy source code with the plugins added.
+
+  Example:
+  ```nix
+  services.caddy = {
+    enable = true;
+    package = pkgs.caddy.withPlugins {
+      plugins = [ "github.com/caddy-dns/powerdns@v1.0.1" ];
+      hash = "sha256-F/jqR4iEsklJFycTjSaW8B/V3iTGqqGOzwYBUXxRKrc=";
+    };
+  };
+  ```
+
+  To get the necessary hash of the vendored dependencies, omit `hash`. The build will fail and tell you the correct value.
+
 - `programs.fzf.keybindings` now supports the fish shell.
 
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->

--- a/pkgs/by-name/ca/caddy/package.nix
+++ b/pkgs/by-name/ca/caddy/package.nix
@@ -6,6 +6,10 @@
 , testers
 , installShellFiles
 , stdenv
+, go
+, xcaddy
+, cacert
+, git
 }:
 let
   version = "2.8.4";
@@ -32,7 +36,8 @@ buildGoModule {
   subPackages = [ "cmd/caddy" ];
 
   ldflags = [
-    "-s" "-w"
+    "-s"
+    "-w"
     "-X github.com/caddyserver/caddy/v2.CustomVersion=${version}"
   ];
 
@@ -61,12 +66,83 @@ buildGoModule {
       --zsh <($out/bin/caddy completion zsh)
   '';
 
-  passthru.tests = {
-    inherit (nixosTests) caddy;
-    version = testers.testVersion {
-      command = "${caddy}/bin/caddy version";
-      package = caddy;
+  passthru = {
+    tests = {
+      inherit (nixosTests) caddy;
+      version = testers.testVersion {
+        command = "${caddy}/bin/caddy version";
+        package = caddy;
+      };
     };
+    withPlugins =
+      { plugins
+      , hash ? lib.fakeHash
+      }: caddy.overrideAttrs (finalAttrs: prevAttrs:
+      let
+        pluginsSorted = builtins.sort builtins.lessThan plugins;
+        pluginsList = lib.concatMapStrings (plugin: "${plugin}-") pluginsSorted;
+        pluginsHash = builtins.hashString "md5" pluginsList;
+        pluginsWithoutVersion = builtins.filter (p: !lib.hasInfix "@" p) pluginsSorted;
+      in
+      assert lib.assertMsg (builtins.length pluginsWithoutVersion == 0)
+        "All plugins should have a version (eg ${builtins.elemAt pluginsWithoutVersion 0}@x.y.z)!";
+      {
+        vendorHash = null;
+        subPackages = [ "." ];
+
+        src = stdenv.mkDerivation {
+          pname = "caddy-src-with-plugins-${pluginsHash}";
+          version = finalAttrs.version;
+
+          nativeBuildInputs = [
+            go
+            xcaddy
+            cacert
+            git
+          ];
+          dontUnpack = true;
+          buildPhase =
+            let
+              withArgs = lib.concatMapStrings (plugin: "--with ${plugin} ") pluginsSorted;
+            in
+            ''
+              export GOCACHE=$TMPDIR/go-cache
+              export GOPATH="$TMPDIR/go"
+              XCADDY_SKIP_BUILD=1 TMPDIR="$PWD" xcaddy build v${finalAttrs.version} ${withArgs}
+              (cd buildenv* && go mod vendor)
+            '';
+          installPhase = ''
+            mv buildenv* $out
+          '';
+
+          outputHashMode = "recursive";
+          outputHash = hash;
+          outputHashAlgo = "sha256";
+        };
+
+
+        doInstallCheck = true;
+        installCheckPhase = ''
+          runHook preInstallCheck
+
+          ${lib.toShellVar "notfound" pluginsSorted}
+          while read kind module version; do
+            [[ "$kind" = "dep" ]] || continue
+            module="''${module}@''${version}"
+            for i in "''${!notfound[@]}"; do
+              if [[ ''${notfound[i]} = ''${module} ]]; then
+                unset 'notfound[i]'
+              fi
+            done
+          done < <($out/bin/caddy build-info)
+          if (( ''${#notfound[@]} )); then
+            >&2 echo "Plugins not found: ''${notfound[@]}"
+            exit 1
+          fi
+
+          runHook postInstallCheck
+        '';
+      });
   };
 
   meta = with lib; {

--- a/pkgs/by-name/ca/caddy/package.nix
+++ b/pkgs/by-name/ca/caddy/package.nix
@@ -1,5 +1,6 @@
 { lib
 , buildGoModule
+, callPackage
 , fetchFromGitHub
 , nixosTests
 , caddy
@@ -74,75 +75,7 @@ buildGoModule {
         package = caddy;
       };
     };
-    withPlugins =
-      { plugins
-      , hash ? lib.fakeHash
-      }: caddy.overrideAttrs (finalAttrs: prevAttrs:
-      let
-        pluginsSorted = builtins.sort builtins.lessThan plugins;
-        pluginsList = lib.concatMapStrings (plugin: "${plugin}-") pluginsSorted;
-        pluginsHash = builtins.hashString "md5" pluginsList;
-        pluginsWithoutVersion = builtins.filter (p: !lib.hasInfix "@" p) pluginsSorted;
-      in
-      assert lib.assertMsg (builtins.length pluginsWithoutVersion == 0)
-        "All plugins should have a version (eg ${builtins.elemAt pluginsWithoutVersion 0}@x.y.z)!";
-      {
-        vendorHash = null;
-        subPackages = [ "." ];
-
-        src = stdenv.mkDerivation {
-          pname = "caddy-src-with-plugins-${pluginsHash}";
-          version = finalAttrs.version;
-
-          nativeBuildInputs = [
-            go
-            xcaddy
-            cacert
-            git
-          ];
-          dontUnpack = true;
-          buildPhase =
-            let
-              withArgs = lib.concatMapStrings (plugin: "--with ${plugin} ") pluginsSorted;
-            in
-            ''
-              export GOCACHE=$TMPDIR/go-cache
-              export GOPATH="$TMPDIR/go"
-              XCADDY_SKIP_BUILD=1 TMPDIR="$PWD" xcaddy build v${finalAttrs.version} ${withArgs}
-              (cd buildenv* && go mod vendor)
-            '';
-          installPhase = ''
-            mv buildenv* $out
-          '';
-
-          outputHashMode = "recursive";
-          outputHash = hash;
-          outputHashAlgo = "sha256";
-        };
-
-
-        doInstallCheck = true;
-        installCheckPhase = ''
-          runHook preInstallCheck
-
-          ${lib.toShellVar "notfound" pluginsSorted}
-          while read kind module version; do
-            [[ "$kind" = "dep" ]] || continue
-            module="''${module}@''${version}"
-            for i in "''${!notfound[@]}"; do
-              if [[ ''${notfound[i]} = ''${module} ]]; then
-                unset 'notfound[i]'
-              fi
-            done
-          done < <($out/bin/caddy build-info)
-          if (( ''${#notfound[@]} )); then
-            >&2 echo "Plugins not found: ''${notfound[@]}"
-            exit 1
-          fi
-
-          runHook postInstallCheck
-        '';
-      });
+    withPlugins = callPackage ./plugins.nix { inherit caddy; };
   };
 
   meta = with lib; {

--- a/pkgs/by-name/ca/caddy/plugins.nix
+++ b/pkgs/by-name/ca/caddy/plugins.nix
@@ -1,0 +1,80 @@
+{
+  lib,
+  stdenv,
+  go,
+  xcaddy,
+  cacert,
+  git,
+  caddy,
+}:
+{
+  plugins,
+  hash ? lib.fakeHash,
+}:
+let
+  pluginsSorted = builtins.sort builtins.lessThan plugins;
+  pluginsList = lib.concatMapStrings (plugin: "${plugin}-") pluginsSorted;
+  pluginsHash = builtins.hashString "md5" pluginsList;
+  pluginsWithoutVersion = builtins.filter (p: !lib.hasInfix "@" p) pluginsSorted;
+in
+assert lib.assertMsg (
+  builtins.length pluginsWithoutVersion == 0
+) "All plugins should have a version (eg ${builtins.elemAt pluginsWithoutVersion 0}@x.y.z)!";
+caddy.overrideAttrs (
+  finalAttrs: prevAttrs: {
+    vendorHash = null;
+    subPackages = [ "." ];
+
+    src = stdenv.mkDerivation {
+      pname = "caddy-src-with-plugins-${pluginsHash}";
+      version = finalAttrs.version;
+
+      nativeBuildInputs = [
+        go
+        xcaddy
+        cacert
+        git
+      ];
+      dontUnpack = true;
+      buildPhase =
+        let
+          withArgs = lib.concatMapStrings (plugin: "--with ${plugin} ") pluginsSorted;
+        in
+        ''
+          export GOCACHE=$TMPDIR/go-cache
+          export GOPATH="$TMPDIR/go"
+          XCADDY_SKIP_BUILD=1 TMPDIR="$PWD" xcaddy build v${finalAttrs.version} ${withArgs}
+          (cd buildenv* && go mod vendor)
+        '';
+      installPhase = ''
+        mv buildenv* $out
+      '';
+
+      outputHashMode = "recursive";
+      outputHash = hash;
+      outputHashAlgo = "sha256";
+    };
+
+    doInstallCheck = true;
+    installCheckPhase = ''
+      runHook preInstallCheck
+
+      ${lib.toShellVar "notfound" pluginsSorted}
+      while read kind module version; do
+        [[ "$kind" = "dep" ]] || continue
+        module="''${module}@''${version}"
+        for i in "''${!notfound[@]}"; do
+          if [[ ''${notfound[i]} = ''${module} ]]; then
+            unset 'notfound[i]'
+          fi
+        done
+      done < <($out/bin/caddy build-info)
+      if (( ''${#notfound[@]} )); then
+        >&2 echo "Plugins not found: ''${notfound[@]}"
+        exit 1
+      fi
+
+      runHook postInstallCheck
+    '';
+  }
+)

--- a/pkgs/by-name/ca/caddy/plugins.nix
+++ b/pkgs/by-name/ca/caddy/plugins.nix
@@ -12,14 +12,14 @@
   hash ? lib.fakeHash,
 }:
 let
-  pluginsSorted = builtins.sort builtins.lessThan plugins;
+  pluginsSorted = lib.sort lib.lessThan plugins;
   pluginsList = lib.concatMapStrings (plugin: "${plugin}-") pluginsSorted;
   pluginsHash = builtins.hashString "md5" pluginsList;
-  pluginsWithoutVersion = builtins.filter (p: !lib.hasInfix "@" p) pluginsSorted;
+  pluginsWithoutVersion = lib.filter (p: !lib.hasInfix "@" p) pluginsSorted;
 in
 assert lib.assertMsg (
-  builtins.length pluginsWithoutVersion == 0
-) "All plugins should have a version (eg ${builtins.elemAt pluginsWithoutVersion 0}@x.y.z)!";
+  lib.length pluginsWithoutVersion == 0
+) "All plugins should have a version (eg ${lib.elemAt pluginsWithoutVersion 0}@x.y.z)!";
 caddy.overrideAttrs (
   finalAttrs: prevAttrs: {
     vendorHash = null;

